crypto/src/lib.rs

```diff
@@ -210,14 +210,52 @@ impl KeyPair {
         self.private_key.digest_function()
     }
 
-    /// Construct `KeyPair`
-    pub fn new(public_key: PublicKey, private_key: PrivateKey) -> Self {
+    /// Construct `KeyPair` from a matching pair of public and private key.
+    /// It is up to the user to ensure that the given keys indeed make a pair.
+    #[cfg(not(feature = "std"))]
+    pub fn new_unchecked(public_key: PublicKey, private_key: PrivateKey) -> Self {
         Self {
             public_key,
             private_key,
         }
     }
 
+    /// Construct `KeyPair`
+    ///
+    /// # Errors
+    /// If public and private key don't match, i.e. if they don't make a pair
+    #[cfg(feature = "std")]
+    pub fn new(public_key: PublicKey, private_key: PrivateKey) -> Result<Self, Error> {
+        if public_key.digest_function() != private_key.digest_function() {
+            return Err(Error::KeyGen(String::from("Key pair mismatch")));
+        }
+
+        let algorithm = public_key.digest_function();
+        if algorithm == Algorithm::Ed25519 || algorithm == Algorithm::Secp256k1 {
+            let key_pair = Self::generate_with_configuration(KeyGenConfiguration {
+                key_gen_option: Some(KeyGenOption::FromPrivateKey(private_key)),
+                algorithm,
+            })?;
+
+            if *key_pair.public_key() != public_key {
+                return Err(Error::KeyGen(String::from("Key pair mismatch")));
+            }
+
+            return Ok(key_pair);
+        }
```

Comment:
Is this check with generating public key from secret key is possible only with this algorithms? Is it unavailable for BLS?

Reply:
If so maybe it's better to do match here, so that when a new encryption algorithm will be added this will fail to compile without additional branch?

Reply:
sure, why not